src/shims/fs.rs

```diff
@@ -13,23 +13,58 @@ use helpers::immty_from_uint_checked;
 use shims::time::system_time_to_duration;
 
 #[derive(Debug)]
-pub struct FileHandle {
-    file: File,
-    writable: bool,
+pub enum FileHandle {
+    StdInPlaceholder,
+    StdOutPlaceholder,
+    StdErrPlaceholder,
```

Comment:
I can see the where you are coming from, but is this really worth it? Given that you do `std::cmp::max(min_fd, 3)` in `insert_fd_with_min_fd`, actually having entries in the map for these isn't worth all the extra complication, IMO.

I would not say this if this code here also handled actually printing to / reading from stdin/stdout/stderr, then it would all make sense (and you could even implement `dup` for that). But that is still handled in `foreign_items.rs`.

So I think this should either go all the way and actually implement the standard FDs here, or else the old `struct FileHande`was cleaner overall.

Reply:
Oh and if you decide to move standard FD handling here, I think that's great! It could remove a bunch of annoying hacks. But I think that should be a PR separate from adding new functionality, to keep review simpler.

Reply:
OK, backed it out. Reverting 586c0920646fb54a7b2d9e98962d114550ad6a73 will be a good place to start for future follow-up work.